tests/Operation/WatchFunctionalTest.php

```diff
@@ -1404,6 +1405,67 @@ public function testGetResumeTokenReturnsOriginalResumeTokenOnEmptyBatch()
         $this->assertSame($resumeToken, $changeStream->getResumeToken());
     }
 
+    /**
+     * For a ChangeStream under these conditions:
+     *  - The batch is not empty.
+     *  - The batch hasn’t been iterated at all.
+     *  - Only the initial aggregate command has been executed.
+     * Expected result:
+     *  - getResumeToken must return startAfter from the initial aggregate if the option was specified.
+     *  - getResumeToken must return resumeAfter from the initial aggregate if the option was specified.
+     *  - If neither the startAfter nor resumeAfter options were specified, the getResumeToken result must be empty.
+     */
+    public function testResumeTokenBehaviour()
+    {
+        if (version_compare($this->getServerVersion(), '4.1.1', '<')) {
+            $this->markTestSkipped('Testing resumeAfter and startAfter can only be tested on servers >= 4.1.1');
+        }
+
+        $operation = new Watch($this->manager, $this->getDatabaseName(), $this->getCollectionName(), [], $this->defaultOptions);
+
+        $lastOpTime = null;
+
+        $changeStream = null;
+        (new CommandObserver())->observe(function () use ($operation, &$changeStream) {
+            $changeStream = $operation->execute($this->getPrimaryServer());
+        }, function ($event) use (&$lastOpTime) {
+            $this->assertInstanceOf(CommandSucceededEvent::class, $event['succeeded']);
+            $reply = $event['succeeded']->getReply();
+
+            $this->assertObjectHasAttribute('operationTime', $reply);
+            $lastOpTime = $reply->operationTime;
+        });
+
+        $this->insertDocument(['x' => 1]);
+
+        $changeStream->next();
```

Comment:
I think this should use `advanceCursorUntilValid()` from #659, since this test will also run on sharded clusters.

Reply:
I was going to merge this before #659 since I'm still working on figuring out the cause for the double free. I'd like to go ahead with the merge then change this while rebasing #659.

Reply:
SGTM.